go/ms/main.go

```diff
@@ -0,0 +1,187 @@
+func main() {
+	os.Exit(realMain())
+}
+
+func realMain() int {
+	fatal.Init()
+	env.AddFlags()
+	flag.Parse()
+	if v, ok := env.CheckFlags(&cfg); !ok {
+		return v
+	}
+	if err := setupBasic(); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		return 1
+	}
+	defer log.Flush()
+	defer env.LogAppStopped("MS", cfg.Ms.ID)
+	defer log.HandlePanic()
+	if err := validateConfig(); err != nil {
+		log.Error("Configuration validation failed", "err", err)
+		return 1
+	}
+
+	cfg.Metrics.StartPrometheus()
+	_, err := setupTopo()
+	if err != nil {
+		log.Error("MS setupTopo failed", "err", err)
+		return 1
+	}
+	if err := mscmn.Init(cfg.Ms, cfg.Sciond, cfg.Features); err != nil {
+		log.Error("MS common initialization failed", "err", err)
+		return 1
+	}
+
+	go func() {
+		defer log.HandlePanic()
+		msmsgr.Msgr.ListenAndServe()
+	}()
+
+	if err := setupDb(); err != nil {
+		log.Error("MS db initialization failed", "err", err)
+		return 1
+	}
+
+	defer msmsgr.Msgr.CloseServer()
+	// Start HTTP endpoints.
+	statusPages := service.StatusPages{
+		"info":   service.NewInfoHandler(),
+		"config": service.NewConfigHandler(cfg),
+	}
+	if err := statusPages.Register(http.DefaultServeMux, cfg.Ms.ID); err != nil {
+		log.Error("registering status pages", "err", err)
+		return 1
+	}
+
+	select {
+	case <-fatal.ShutdownChan():
+		return 0
+	case <-fatal.FatalChan():
+		return 1
+	}
+}
+
+func setupDb() error {
+	err := sqlite.New(cfg.Ms.Db, 1)
+	if err != nil {
+		return serrors.WrapStr("setting up database", err)
+	}
+	return nil
+}
+
+func setupTopo() (*ifstate.Interfaces, error) {
+
+	topo, err := topology.FromJSONFile(cfg.General.Topology())
+	if err != nil {
+		return nil, serrors.WrapStr("loading topology", err)
+	}
+
+	intfs := ifstate.NewInterfaces(topo.IFInfoMap(), ifstate.Config{})
+	itopo.Init(&itopo.Config{
+		ID:  cfg.General.ID,
+		Svc: proto.ServiceType_ms,
+		Callbacks: itopo.Callbacks{
+			OnUpdate: func() {
+				intfs.Update(itopo.Get().IFInfoMap())
+			},
+		},
+	})
+
+	if err := itopo.Update(topo); err != nil {
+		return nil, serrors.WrapStr("setting initial static topology", err)
+	}
+	infraenv.InitInfraEnvironment(cfg.General.Topology())
+	return intfs, nil
+}
+
+// setupBasic loads the config from file and initializes logging.
+func setupBasic() error {
+	// Load and initialize config.
+	md, err := toml.DecodeFile(env.ConfigFile(), &cfg)
+	if err != nil {
+		return serrors.WrapStr("Failed to load config", err, "file", env.ConfigFile())
+	}
+	if len(md.Undecoded()) > 0 {
+		return serrors.New("Failed to load config: undecoded keys", "undecoded", md.Undecoded())
+	}
+	cfg.InitDefaults()
+	if err := log.Setup(cfg.Logging); err != nil {
+		return serrors.WrapStr("Failed to initialize logging", err)
+	}
+	prom.ExportElementID(cfg.Ms.ID)
+	return env.LogAppStarted("MS", cfg.Ms.ID)
+}
+
+func validateConfig() error {
+	if err := cfg.Validate(); err != nil {
+		return err
+	}
+	if cfg.Metrics.Prometheus == "" {
+		cfg.Metrics.Prometheus = "127.0.0.1:1285"
```

Comment:
`"127.0.0.1:1285"`

Defaults should be documented.

For the CS, the default is to not export metrics if the prometheus address is not configured.

Reply:
I changed it to export it only if the address is configured